gluten-core/src/main/scala/io/glutenproject/extension/ExpandFallbackPolicy.scala

```diff
@@ -0,0 +1,164 @@
+        // we plan exchange to columnar exchange in columnar rules and the exchange does not
+        // support columnar, so the output columnar is always false in AQE postStageCreationRules
+        case ColumnarToRowExec(s: Exchange) if s.supportsColumnar && isAdaptiveContext =>
+          countFallback(s)
+        case ColumnarToRowExec(p: GlutenPlan) =>
+          logDebug(s"Find a columnar to row for gluten plan:\n$p")
+          fallbacks = fallbacks + 1
+          countFallback(p)
+        case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
+          // Possible fallback for leaf node.
+          fallbacks = fallbacks + 1
+        case p => p.children.foreach(countFallback)
+      }
+    }
+    countFallback(plan)
+    fallbacks
+  }
+
+  private def hasColumnarBroadcastExchangeWithJoin(plan: SparkPlan): Boolean = {
+    def isColumnarBroadcastExchange(p: SparkPlan): Boolean = p match {
+      case BroadcastQueryStageExec(_, _: ColumnarBroadcastExchangeExec, _) => true
+      case _ => false
+    }
+
+    plan.find {
+      case j: BroadcastHashJoinExecTransformer
+          if isColumnarBroadcastExchange(j.left) ||
+            isColumnarBroadcastExchange(j.right) =>
+        true
+      case _ => false
+    }.isDefined
+  }
+
+  private def fallback(plan: SparkPlan): Option[String] = {
+    val fallbackThreshold = if (isAdaptiveContext) {
+      GlutenConfig.getConf.wholeStageFallbackThreshold
+    } else if (plan.find(_.isInstanceOf[AdaptiveSparkPlanExec]).isDefined) {
+      // if we are here, that means we are now at `QueryExecution.preparations` and
+      // AQE is actually applied. We do nothing for this case, and later in
+      // AQE we can check `wholeStageFallbackThreshold`.
+      return None
+    } else {
+      // AQE is not applied, so we use the whole query threshold to check if should fallback
+      GlutenConfig.getConf.queryFallbackThreshold
+    }
+    if (fallbackThreshold < 0) {
+      return None
+    }
+
+    // not safe to fallback row-based BHJ as the broadcast exchange is already columnar
+    if (hasColumnarBroadcastExchangeWithJoin(plan)) {
+      return None
+    }
+
+    val numFallback = countFallbacks(plan)
+    if (numFallback >= fallbackThreshold) {
+      Some(
+        s"Fall back the plan due to fallback number $numFallback, " +
+          s"threshold $fallbackThreshold")
+    } else {
+      None
+    }
+  }
+
+  private def fallbackToRowBasedPlan(): SparkPlan = {
+    val transformPostOverrides = TransformPostOverrides(isAdaptiveContext)
+    val planWithColumnarToRow = InsertTransitions.insertTransitions(originalPlan, false)
+    planWithColumnarToRow.transform {
+      case c2r @ ColumnarToRowExec(_: ShuffleQueryStageExec) =>
+        transformPostOverrides.transformColumnarToRowExec(c2r)
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    val reason = fallback(plan)
+    if (reason.isDefined) {
```

Comment:
may worth to add a warning log to indicate the stage/query is running fallback now

Reply:
yeah, it's wroth. Now, it follows the `GlutenFallbackReporter`. It would print log and post event to Spark UI.